patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs

```diff
@@ -0,0 +1,116 @@
+			set => ModSlotPlayer.exHideAccessory[Type] = value;
+		}
+
+		protected sealed override void Register() => Type = LoaderManager.Get<AccessorySlotLoader>().Register(this);
+
+		public bool IsEmpty() => FunctionalItem.IsAir && VanityItem.IsAir && DyeItem.IsAir;
+
+		/// <summary>
+		/// This function allows for advanced custom drawing for the accessory slot. Called for Dyes, Vanity, and Functionals.
+		/// By default runs ItemSlot.Draw()
+		/// Receives data:
+		/// <para><paramref name="inv"/> :: the array containing all accessory slots, yours is inv[slot] </para>
+		/// <para><paramref name="slot"/> :: which is the index for inventory that you were assigned </para>
+		/// <para><paramref name="position"/> :: is the position of where the ItemSlot will be drawn </para>
+		/// <para><paramref name="context"/> :: 12 => dye; 11 => vanity; 10 => functional </para>
+		/// </summary>
+		public virtual void DrawModded(Item[] inv, int context, int slot, Vector2 position) =>
+			ItemSlot.Draw(Main.spriteBatch, inv, context, slot, position);
+
+		/// <summary>
+		/// Override to set conditions on when the slot is valid for stat/vanity calculations and player usage.
+		/// Example: the demonHeart is consumed and in Expert mode in Vanilla.
+		/// </summary>
+		public virtual bool IsEnabled() => true;
+
+		/// <summary>
+		/// Override to replace the vanilla effect behavior of the slot with your own.
+		/// By default calls:
+		/// Player.VanillaUpdateEquips(FunctionalItem), Player.ApplyEquipFunctional(FunctionalItem, ShowVisuals), Player.ApplyEquipVanity(VanityItem)
+		/// </summary>
+		public virtual void ApplyEquipEffects() {
+			Player.VanillaUpdateEquip(FunctionalItem);
+			Player.ApplyEquipFunctional(FunctionalItem, ShowVisuals);
+			Player.ApplyEquipVanity(VanityItem);
+		}
+
+		/// <summary>
+		/// Override to set conditions on what can be placed in the slot. Return false to prevent the item going in slot. Return true for dyes, if you want dyes. Example: only wings can go in slot.
+		/// Receives data:
+		/// <para><paramref name="checkItem"/> :: the item that is attempting to enter the slot </para>
+		/// </summary>
+		public virtual bool CanAcceptItem(Item checkItem) => true;
+
+		/// <summary>
+		/// Override to change the condition on when the slot is visible, but otherwise non-functional for stat/vanity calculations.
+		/// Defaults to check 'property' MySlotContainsAnItem()
+		/// </summary>
+		/// <returns></returns>
+		public virtual bool IsVisibleWhenNotEnabled() => !IsEmpty();
+
+		/// <summary>
+		/// After checking for empty slots in ItemSlot.AccessorySwap, this allows for changing what the default target slot (accSlotToSwapTo) will be.
+		/// DOES NOT affect vanilla behaviour of swapping items like for like where existing in a slot
+		/// Return true to set this slot as the default targetted slot.
+		/// </summary>
+		public virtual bool ModifyDefaultSwapSlot(Item item, int accSlotToSwapTo) => false;
```

Comment:
I personally need a hook for when an item is quick swapped into my slot.

Reply:
An OnEnterSlot(context) hook for all movement vectors or just from quick swap?

Reply:
A hook seems unnecessary - why not just track on the AccessorySlot itself when FunctionalItem.Type changes?

Is it specifically that are looking for the context as to how it got in to the slot?

Reply:
A hook seems unnecessary - why not just track on the AccessorySlot itself when FunctionalItem.Type changes?

Is it specifically that are looking for the context as to how it got in to the slot?